docs/developer/logging-guidelines.md

```diff
@@ -0,0 +1,407 @@
+[here](https://github.com/uber-go/zap/blob/master/FAQ.md#why-sample-application-logs).
+
+[sampling]:https://github.com/kubernetes-sigs/controller-runtime/blob/b1d6919d3e12fa85a119dd9792bdfdc17bdf8c3b/pkg/log/zap/zap.go#L211-L215
+
+### Logger Dependency in Code
+
+#### Logger Initialization
+
+The logger should be initialized in the `main` package and passed to the components as `logr.Logger`. In the `main`
+package
+we set the parameters of the logger, such as the level, stack trace level, and format. The control-plane components
+should
+not change them.
+
+#### Logger Injection
+
+For asynchronous components, inject the logger during the component creation, setting up the name of the logger to
+match the name of the component:
+
+```go
+asyncComp := newComponent(logger.WithName("async-component"))
+```
+
+For synchronous components (meaning another component will call it synchronously), inject the logger during the
+method call, adding extra context when necessary:
+
+```go
+l := logger.WithName("sync-component").WithValues("operationID", 1)
+syncComp.Process(l)
+```
+
+#### Special Case - Reconciler
+
+The Reconciler [gets its logger][reconciler-logger] from the controller-runtime.
+That logger is compatible with our logging guidelines.
+
+[reconciler-logger]:https://github.com/nginxinc/nginx-kubernetes-gateway/blob/5547fe5472d1742a937c8adbbd399893ee30f9e1/internal/framework/controller/reconciler.go#L63
+
+#### Unit Tests
+
+In unit tests, we don't test what we log.
+
+To initialize a logger, use `zap.New()` from [controller-runtime helpers][cr-zap].
+
+### External Libraries
+
+There are two critical libraries for NKG that log:
+
+- [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime).
+  - The [log levels][cr-log-levels] are compatible with our project.
+  - We [inject][inject] logger into the library as `logr.Logger` from [logr][logr].
+  - See also
+      the [logging guidelines][cr-logging-guidelines]
+      of that project.
+- [client-go](https://github.com/kubernetes/client-go).
+  - It uses [klog](https://github.com/kubernetes/klog) for logging.
+  - Most of the logging is done at increased klog-specific
+      verbosity. However, errors are logged at the default verbosity like
+      in [this line](https://github.com/kubernetes/client-go/blob/c5b1c13ccbedeb03c00ba162ef27566b0dfb512d/tools/record/event.go#L240).
+
+[inject]:https://github.com/nginxinc/nginx-kubernetes-gateway/blob/9b3ae2c7c59f28213a7690e049d9996443dbd3fc/internal/mode/static/manager.go#L54
+
+[cr-logging-guidelines]:https://github.com/kubernetes-sigs/controller-runtime/blob/b1d6919d3e12fa85a119dd9792bdfdc17bdf8c3b/TMP-LOGGING.md
+
+When adding a new library, evaluate if it does logging and how. If it is possible, configure the library logging to
+be compatible with NKG logging. If not, document the logs in the user doc, so that the users are prepared for them.
+
+### Evolution
```

Comment:
do we anticipate that our logging will drastically change in the future? if so, will our current choices and current guidelines and design, allow us to address new requirements in the future easily?

Reply:
I don't think our logging approach will change drastically, however, we may have a requirement to add support for exporting logs. 

Have you taken a look at the OTEL logging documentation https://opentelemetry.io/docs/specs/otel/logs/?

Will our tooling and guidelines be compatible with this?

Reply:
Makes sense. Thanks for digging into this